node/core/dispute-coordinator/src/real/ordering/mod.rs

```diff
@@ -259,41 +272,51 @@ async fn get_block_number(
 	send_message_fatal(sender, ChainApiMessage::BlockNumber(relay_parent, tx), rx).await
 }
 
-async fn get_finalized_block_hash(sender: &mut impl SubsystemSender) -> FatalResult<Option<Hash>> {
+async fn get_finalized_block(
+	sender: &mut impl SubsystemSender,
+) -> FatalResult<Option<(BlockNumber, Hash)>> {
 	let (number_tx, number_rx) = oneshot::channel();
 	let number =
 		send_message_fatal(sender, ChainApiMessage::FinalizedBlockNumber(number_tx), number_rx)
 			.await?;
 
 	let (hash_tx, hash_rx) = oneshot::channel();
-	send_message_fatal(sender, ChainApiMessage::FinalizedBlockHash(number, hash_tx), hash_rx).await
+	let maybe_hash =
+		send_message_fatal(sender, ChainApiMessage::FinalizedBlockHash(number, hash_tx), hash_rx)
+			.await?;
+
+	Ok(Some(number).zip(maybe_hash))
 }
 
-/// Returns ancestor hashes of `head` in the descending order, stopping
-/// either at `last_observed_hash` or the genesis block.
+/// Returns ancestors of `head` in the descending order, stopping
+/// either at `last_observed_hash` or the latest  finalized block.
 ///
-/// Suited specifically for querying non-finalized chains, thus doesn't
-/// rely on block numbers.
+/// Suited specifically for querying non-finalized chains, thus
+/// doesn't rely on block numbers.
 ///
-/// The caller must ensure that `last_observed_hash` is indeed an ancestor
-/// of `head` in order to avoid traversing the whole chain.
 /// Both `head` and `last_observed_hash` are **not** included in the result.
-async fn get_block_ancestors_hashes<Sender>(
+async fn get_block_ancestors<Sender>(
 	sender: &mut Sender,
 	mut head: Hash,
 	last_observed_hash: Hash,
-) -> NonFatalResult<Vec<Hash>>
+) -> NonFatalResult<Vec<(BlockNumber, Hash)>>
 where
 	Sender: SubsystemSender,
 {
-	const ANCESTRY_STEP: usize = 4;
+	const ANCESTRY_STEP: usize = 10;
 
 	let mut ancestors = Vec::new();
 
 	if head == last_observed_hash {
 		return Ok(ancestors)
 	}
 
+	let (_, finalized_block_hash) = if let Ok(Some(hash)) = get_finalized_block(sender).await {
+		hash
```

Comment:
Ok great - we have a safeguard. I would suggest to use the last finalized block number here, as it seems to be retrievable more reliably than the hash (no `Option`), then we can also avoid that if/else which makes me a bit uneasy as it would mean, that in case of a hash lookup error, we would prioritize finalized candidates, which is not great.